boot/freeldr/freeldr/ntldr/winldr.c

```diff
@@ -496,6 +493,69 @@ LoadModule(
     return Success;
 }
 
+static
+BOOLEAN
+WinLdrIsPaeSupported(IN PLOADER_PARAMETER_BLOCK LoaderBlock,
+                     IN PCSTR BootOptions,
+                     IN USHORT OperatingSystemVersion,
+                     IN PCHAR KernelFileName,
+                     IN PCHAR HalFileName)
+{
+    BOOLEAN PaeEnabled = FALSE;
+    BOOLEAN PaeDisabled = FALSE;
+    BOOLEAN NoexecuteDisabled = FALSE;
+    BOOLEAN NoexecuteEnabled = FALSE;
+    BOOLEAN Result;
+
+    UNIMPLEMENTED;
+
+    if (OperatingSystemVersion > _WIN32_WINNT_NT4 &&
+        NtLdrGetOption(BootOptions, "PAE"))
+    {
+        /* We found the PAE option. */
+        PaeEnabled = TRUE;
+    }
+
+    Result = PaeEnabled;
+
+    if (OperatingSystemVersion > _WIN32_WINNT_WIN2K)
+    {
+        if (NtLdrGetOption(BootOptions, "NOPAE"))
+            PaeDisabled = TRUE;
+
+        if (!LoaderBlock->SetupLdrBlock)
+        {
+            if (NtLdrGetOption(BootOptions, "NOEXECUTE=ALWAYSOFF"))
+                NoexecuteDisabled = TRUE;
+            else if (NtLdrGetOption(BootOptions, "NOEXECUTE"))
+                NoexecuteEnabled = TRUE;
+
+            if (NtLdrGetOption(BootOptions, "EXECUTE") && !NoexecuteEnabled)
+                NoexecuteDisabled = TRUE;
+        }
+    }
+
+    if (SafeBoot)
+    {
+        PaeDisabled = TRUE;
+        NoexecuteDisabled = TRUE;
+    }
+
+    TRACE("NoexecuteDisabled %X\n", NoexecuteDisabled);
+    TRACE("PaeEnabled %X, PaeDisabled %X\n", PaeEnabled, PaeDisabled);
+
+    if (PaeDisabled)
+        Result = FALSE;
+
+    if (NoexecuteDisabled == FALSE)
```

Comment:
So this will override `PaeDisabled`? The logic seems to be a bit broken here.
I don't think NOEXECUTE should be parsed in this function - it depends on PAE, but can be both turned off even with PAE enabled, isn't it?